### PR TITLE
[tmva][sofie] Improve AD-friendlieness of emitted code for Clad

### PR DESCRIPTION
**Pull request by guitargeek**
This commit refactors SOFIE-generated inference code to enable correct and efficient reverse-mode automatic differentiation with Clad.

Key changes:

* Introduce explicit primitive operations (`Copy`, `Fill`, `Relu`) in SOFIE_common.hxx and provide corresponding custom pullbacks in CladDerivator.h. This replaces previously inlined loops and allows Clad to generate efficient gradient code without relying on tapes or loop-level differentiation.

* Update Gemm code generation to emit Copy/Fill instead of manually expanding bias initialization loops. This better exposes the intent and improves AD performance and correctness.

* Replace manual ReLU loops with a dedicated Relu() call, enabling a custom pullback that avoids tape-based condition tracking.

* Generate an additional "unoptimized" model variant in the SOFIE test suite (`OptimizationLevel::kBasic`), and use it for AD tests. This disables memory reuse of intermediate tensors. Opaque memory reuse is safe for inference but breaks source-transformation AD.

* Improve gradient test diagnostics in SOFIE Clad tests by reporting mismatched indices instead of only checking a global max difference.

With these changes, Clad-generated gradients for SOFIE models are both correct and significantly faster, reaching performance comparable to frameworks such as PyTorch and JAX on the CPU for the tested cases (fully-connected neural networks with multiple layers).

### CODE REVIEW

File: tmva/sofie/inc/TMVA/SOFIE_common.hxx [hunk 1]
```diff
@@ -772,6 +772,27 @@ inline void Gemm_Call(float *output, bool transa, bool transb, int m, int n, int
                                            &beta, output, ldc);
 }
 
+inline void Fill(float *output, float value, int size)
+{
+   for (int i = 0; i < size; i++) {
+      output[i] = value;
+   }
+}
+
```
Comment (lmoneta) [inline tmva/sofie/inc/TMVA/SOFIE_common.hxx:779]: Is it better for AD here to have an implementation or use std::copy?

**Review by lmoneta — APPROVED** [review]
LGTM!
Thank you, Jonas, for these improvements, which help CLAD. I have only a question on whether is better to impelment a Copy function or using directly std::copy.

File: tmva/sofie/inc/TMVA/SOFIE_common.hxx (hunk 1 above)
Comment (guitargeek) [reply tmva/sofie/inc/TMVA/SOFIE_common.hxx:779]: It doesn't matter because we're defining a custom pullback anyway, we we can use what we like. I'll try `std::copy` :+1:

Comment (guitargeek) [reply tmva/sofie/inc/TMVA/SOFIE_common.hxx:779]: Ah sorry I misunderstood you comment before: we need an implementation to define a custom pullback. But I used `std::fill` and `std::copy` in the implementation now.
